### PR TITLE
Floating point numbers

### PR DESCRIPTION
**Pull request by vakaras**

### CODE REVIEW

File: prusti-common/src/vir/ast/common.rs [hunk 1]
```diff
@@ -193,6 +202,9 @@ impl Type {
             Type::Domain(ref pred_name) => pred_name.to_string(),
             Type::Snapshot(ref pred_name) => pred_name.to_string(),
             Type::Seq(_) => "Seq".to_string(),
+            Type::TypedRef(ref pred_name) => format!("{}", pred_name),
+            Type::Domain(ref pred_name) => format!("{}", pred_name),
+            Type::Float(_) => "float".to_string(),
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/common.rs]: Should this produce different values for different sizes?

File: prusti-common/src/vir/ast/common.rs [hunk 2]
```diff
@@ -317,3 +330,24 @@ impl WithIdentifier for Field {
         self.name.clone()
     }
 }
+
+// maybe missing some other impl
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum BackendFunc {
+    BVfunc(String, String),
+    Floatfunc(String, String)
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/common.rs]:
```suggestion
    BitvecFunc(String, String),
    FloatFunc(String, String)
```
?

File: prusti-common/src/vir/ast/common.rs [hunk 3]
```diff
@@ -317,3 +330,24 @@ impl WithIdentifier for Field {
         self.name.clone()
     }
 }
+
+// maybe missing some other impl
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/common.rs]: Comment still relevant?

File: prusti-common/src/vir/ast/expr.rs [hunk 4]
```diff
@@ -73,6 +75,7 @@ pub enum PlaceComponent {
 pub enum UnaryOpKind {
     Not,
     Minus,
+    IsNaN,
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/expr.rs:76]: Why do we treat this as a unary operator in VIR? Why not e.g. a builtin?

File: prusti-common/src/lib.rs [hunk 5]
```diff
@@ -23,6 +23,8 @@ extern crate serde;
 extern crate uuid;
 extern crate viper;
 
+
+
```
Comment (Aurel300) [inline prusti-common/src/lib.rs]:
```suggestion
```

File: prusti-common/src/vir/ast/expr.rs [hunk 6]
```diff
@@ -99,11 +107,18 @@ pub enum ContainerOpKind {
     // more to follow if required
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub enum FloatConst {
+    FloatConst64(u64),
+    FloatConst32(u32),
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/expr.rs:106]: What is the issue with storing `f64/32` directly? They would end up being the same IEEE representations, no?

File: prusti-common/src/vir/ast/expr.rs [hunk 7]
```diff
@@ -228,6 +243,16 @@ impl fmt::Display for Expr {
                     .collect::<Vec<String>>()
                     .join(", "),
             ),
+            // mimicked DomainFuncApp for Display
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/expr.rs]:
```suggestion
```

File: prusti-common/src/vir/ast/expr.rs [hunk 8]
```diff
@@ -228,6 +243,16 @@ impl fmt::Display for Expr {
...
+            Expr::BackendFuncApp(ref function, ref args, ref _pos) => write!(
+                f,
+                "{}({})",
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/expr.rs]: Does the function name always contain e.g. `backend` or something to identify it somewhat uniquely in string dumps?

File: prusti-common/src/vir/ast/expr.rs [hunk 9]
```diff
@@ -518,6 +573,15 @@ impl Expr {
         Expr::DomainFuncApp(func, args, Position::default())
     }
 
+    // Mimicked domain_func_app above
+    pub fn backend_func_app(
+        func: BackendFunc,
+        args: Vec<Expr>,
+    ) -> Self {
+        Expr::
+        BackendFuncApp(func, args, Position::default())
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/expr.rs]:
```suggestion
        Expr::BackendFuncApp(func, args, Position::default())
```

File: prusti-common/src/vir/ast/expr.rs [hunk 10]
```diff
@@ -1027,10 +1091,15 @@ impl Expr {
             Expr::DomainFuncApp(ref func, _, _) => {
                 &func.return_type
             },
+            Expr::BackendFuncApp(ref _func, _, _) => {
+                unimplemented!()
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/expr.rs]: The type is not something unexpressible, is it?

File: prusti-common/src/vir/ast/expr_transformers.rs [hunk 11]
```diff
@@ -707,7 +727,21 @@ pub trait FallibleExprFolder: Sized {
...
+    ) -> Result<Expr, Self::Error> {
+        Ok(Expr::BackendFuncApp(
+            func,
+            args.into_iter()
+                .map(|e| self.fallible_fold(e))
+                .collect::<Result<Vec<_>, Self::Error>>()?,
+            pos
+        ))
+    }
+    fn fallible_inhale_exhale(
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/expr_transformers.rs]:
```suggestion
    fn fallible_fold_inhale_exhale(
```
Then also undo the change on line 833. It seems a rebase accidentally renamed this function when it should not have.

File: prusti-common/src/vir/ast/function.rs [hunk 12]
```diff
@@ -86,6 +86,7 @@ pub fn compute_identifier(name: &str, formal_args: &[LocalVar], return_type: &Ty
             Type::Domain(ref name) => name.to_string(),
             Type::Snapshot(ref name) => format!("Snap${}", name),
             Type::Seq(ref elem_ty) => format!("Seq${}", type_name(elem_ty)),
+            Type::Float(_) => "$float$".to_string(),
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/function.rs]: Again – different values for `f32/64`?

File: prusti-common/src/vir/to_viper.rs [hunk 13]
```diff
@@ -341,52 +343,72 @@ impl<'v> ToViper<'v, viper::Expr<'v>> for Expr {
...
+                    }
+                },
+
+            _ =>  match op {
+                BinOpKind::Mod => ast.module_with_pos(left.to_viper(ast), right.to_viper(ast), pos.to_viper(ast)),
+                BinOpKind::And => ast.and_with_pos(left.to_viper(ast), right.to_viper(ast), pos.to_viper(ast)),
+                BinOpKind::Or => ast.or_with_pos(left.to_viper(ast), right.to_viper(ast), pos.to_viper(ast)),
+                BinOpKind::Implies => ast.implies_with_pos(left.to_viper(ast), right.to_viper(ast), pos.to_viper(ast)),
+                BinOpKind::BitAnd | BinOpKind::BitOr | BinOpKind::BitXor | BinOpKind::Shl | BinOpKind::Shr => unimplemented!("Bitwise Binops unimplemented"),
+                _ => unimplemented!("Case should be unreachable")
```
Comment (Aurel300) [inline prusti-common/src/vir/to_viper.rs]:
```suggestion
                _ => match op {
                    BinOpKind::Mod => ast.module_with_pos(left.to_viper(ast), right.to_viper(ast), pos.to_viper(ast)),
                    BinOpKind::And => ast.and_with_pos(left.to_viper(ast), right.to_viper(ast), pos.to_viper(ast)),
                    BinOpKind::Or => ast.or_with_pos(left.to_viper(ast), right.to_viper(ast), pos.to_viper(ast)),
                    BinOpKind::Implies => ast.implies_with_pos(left.to_viper(ast), right.to_viper(ast), pos.to_viper(ast)),
                    BinOpKind::BitAnd | BinOpKind::BitOr | BinOpKind::BitXor | BinOpKind::Shl | BinOpKind::Shr => unimplemented!("Bitwise Binops unimplemented"),
                    _ => unreachable!()
```

File: prusti-common/src/vir/to_viper.rs [hunk 14]
```diff
@@ -341,52 +343,72 @@ impl<'v> ToViper<'v, viper::Expr<'v>> for Expr {
...
+                        UnaryOpKind::Minus => UnOpFloat::Neg,
+                        UnaryOpKind::IsNaN => UnOpFloat::IsNan,
+                        _ => unimplemented!("unimplemented unop for floats")
+                    };
+                    ast.float_unop(op_kind, f_size, expr.to_viper(ast))
                 }
-                BinOpKind::Or => {
-                    ast.or_with_pos(left.to_viper(ast), right.to_viper(ast), pos.to_viper(ast))
+
+                _ =>  match op {
```
Comment (Aurel300) [inline prusti-common/src/vir/to_viper.rs]:
```suggestion
                _ => match op {
```

File: prusti-common/src/vir/to_viper.rs [hunk 15]
```diff
@@ -461,6 +483,16 @@ impl<'v> ToViper<'v, viper::Expr<'v>> for Expr {
                     &[], // TODO not necessary so far
                 )
             }
+
+            Expr::BackendFuncApp(ref _function, ref _args, ref _pos) => {
+                unimplemented!("ToViper for BackendFuncApp is unimplemented")
```
Comment (Aurel300) [inline prusti-common/src/vir/to_viper.rs]: Do the tests not use backend functions?

File: prusti-tests/tests/verify/pass/floats/f32vsf64.rs [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    // both work
+
+    // let x = 5.3;
+    // assert!(x == x); 
+    // let v = 213412341.34645634653;
+    // assert!(v - v == 0.);
+    // assert!(v - v <= 0.);
+    // assert!(v - v >= 0.);
+
+    assert!(true);
```
Comment (Aurel300) [inline prusti-tests/tests/verify/pass/floats/f32vsf64.rs]: Not a proper test file…

File: prusti-tests/tests/verify/pass/floats/is_nan_f64.rs [hunk 17]
```diff
@@ -0,0 +1,12 @@
+fn main() {
+
+    // Can't handle, but can for f32
```
Comment (Aurel300) [inline prusti-tests/tests/verify/pass/floats/is_nan_f64.rs:3]: Why not?

File: prusti-tests/tests/verify/pass/floats/predicate_float.rs [hunk 18]
```diff
@@ -0,0 +1,14 @@
+extern crate prusti_contracts;
+use prusti_contracts::*;
+
+#[predicate]
```
Comment (Aurel300) [inline prusti-tests/tests/verify/pass/floats/predicate_float.rs]: This is old syntax (https://github.com/viperproject/prusti-dev/pull/537). Also what is this testing? These functions can simply be `#[pure]`, they are not using Prusti syntax.

File: prusti-viper/src/encoder/builtin_encoder.rs [hunk 19]
```diff
@@ -59,6 +63,7 @@ impl BuiltinEncoder {
             BuiltinMethodKind::HavocBool => "builtin$havoc_bool".to_string(),
             BuiltinMethodKind::HavocInt => "builtin$havoc_int".to_string(),
             BuiltinMethodKind::HavocRef => "builtin$havoc_ref".to_string(),
+            BuiltinMethodKind::HavocF32 | BuiltinMethodKind::HavocF64 => "builtin$havoc_float".to_string(),
```
Comment (Aurel300) [inline prusti-viper/src/encoder/builtin_encoder.rs:66]: Can the same method accept both types? We are not mangling these names, this seems like it should fail.

File: prusti-viper/src/encoder/mir_encoder/mod.rs [hunk 20]
```diff
@@ -484,43 +485,93 @@ impl<'p, 'v: 'p, 'tcx: 'v> MirEncoder<'p, 'v, 'tcx> {
         self.encoder.encode_snapshot_type(ty)
     }
 
+
+    pub fn is_nan_const(e: &vir::Expr) -> bool {
+        match e {
+            vir::Expr::Const(vir::Const::Float(vir::FloatConst::FloatConst64(val)), _) => *val == 9221120237041090560,
+            vir::Expr::Const(vir::Const::Float(vir::FloatConst::FloatConst32(val)), _) => *val == 2143289344,
```
Comment (Aurel300) [inline prusti-viper/src/encoder/mir_encoder/mod.rs]:
```suggestion
            vir::Expr::Const(vir::Const::Float(vir::FloatConst::FloatConst64(val)), _) => *val == 0x7ff8000000000000,
            vir::Expr::Const(vir::Const::Float(vir::FloatConst::FloatConst32(val)), _) => *val == 0x7fc00000,
```

File: prusti-viper/src/encoder/mir_encoder/mod.rs [hunk 21]
```diff
@@ -484,43 +485,93 @@ impl<'p, 'v: 'p, 'tcx: 'v> MirEncoder<'p, 'v, 'tcx> {
...
+            _ => false
+        }
+    }
+
+    pub fn encode_const_float_eq_cmp(&self, left: vir::Expr, right: vir::Expr) -> vir::Expr {
+        if MirEncoder::<'p, 'v, 'tcx>::is_nan_const(&left) || MirEncoder::<'p, 'v, 'tcx>::is_nan_const(&right) {
```
Comment (Aurel300) [inline prusti-viper/src/encoder/mir_encoder/mod.rs]: Are the explicit lifetimes required here?

File: prusti-viper/src/encoder/mir_encoder/mod.rs [hunk 22]
```diff
@@ -484,43 +485,93 @@ impl<'p, 'v: 'p, 'tcx: 'v> MirEncoder<'p, 'v, 'tcx> {
...
-                ))
-            }
-            unsupported_op => {
-                return Err(EncodingError::unsupported(format!(
-                    "operation '{:?}' is not supported",
-                    unsupported_op
-                )))
+        match (&left, &right) {
+            (vir::Expr::Const(vir::Const::Float(_), _), _) 
+            | (_, vir::Expr::Const(vir::Const::Float(_), _))
```
Comment (Aurel300) [inline prusti-viper/src/encoder/mir_encoder/mod.rs]: It's strange that this match is only for float constants in VIR unlike the `bool` match which checks the type. What if I wanted to encode a float operation with two non-constant operands?

File: prusti-viper/src/encoder/mir_encoder/mod.rs [hunk 23]
```diff
@@ -484,43 +485,93 @@ impl<'p, 'v: 'p, 'tcx: 'v> MirEncoder<'p, 'v, 'tcx> {
...
+                            unsupported_op
+                        )))
+                    }
+                })
             }
-        })
+        }
+
+
+
```
Comment (Aurel300) [inline prusti-viper/src/encoder/mir_encoder/mod.rs]:
```suggestion
```

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 24]
```diff
@@ -2144,12 +2149,12 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
...
-                                    term.source_info.span,
-                                    args,
-                                    destination,
-                                    vir::BinOpKind::EqCmp,
+                                def_id,
+                                location,
+                                term.source_info.span,
+                                args,
+                                destination,
+                                vir::BinOpKind::EqCmp,
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]:
```suggestion
                                    def_id,
                                    location,
                                    term.source_info.span,
                                    args,
                                    destination,
                                    vir::BinOpKind::EqCmp,
```

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 25]
```diff
@@ -2164,16 +2169,49 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
...
-                                    term.source_info.span,
-                                    args,
-                                    destination,
-                                    vir::BinOpKind::NeCmp,
+                                def_id,
+                                location,
+                                term.source_info.span,
+                                args,
+                                destination,
+                                vir::BinOpKind::NeCmp,
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]:
```suggestion
                                    def_id,
                                    location,
                                    term.source_info.span,
                                    args,
                                    destination,
                                    vir::BinOpKind::NeCmp,
```

File: prusti-common/src/vir/ast/expr.rs [hunk 26]
```diff
@@ -440,6 +475,26 @@ impl Expr {
...
+    }
+
+    pub fn shl(left: Expr, right: Expr) -> Self {
+        Expr::BinOp(BinOpKind::Shl, box left, box right, Position::default())
+    }
+
+    pub fn shr(left: Expr, right: Expr) -> Self {
+        Expr::BinOp(BinOpKind::Shr, box left, box right, Position::default())
+    }
+
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/expr.rs]: One for `is_nan` as well?

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 27]
```diff
@@ -2164,16 +2169,49 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
...
                                 )?
                             );
                         }
 
+                        "core::f32::<impl f32>::is_nan" |
+                        "core::f64::<impl f64>::is_nan" => {
+                            let span = self.mir_encoder.get_span_of_location(location);
+                            let operand = self.mir_encoder.encode_operand_expr(&args[0]).with_span(span)?;
+                            let operand_box = std::boxed::Box::from(operand);
+                            let expr = vir::Expr::UnaryOp(UnaryOpKind::IsNaN, operand_box, vir::Position::default());
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]: Assuming `Expr::is_nan` was added:
```suggestion
                            let expr = vir::Expr::is_nan(operand);
```

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 28]
```diff
@@ -2164,16 +2169,49 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
...
+                                .unwrap();
+
+                            stmts.push(vir::Stmt::Inhale(
+                                type_predicate,
+                            ));
+
+                            // Store a label for permissions got back from the call
+                            self.label_after_location.insert(location, label.clone());
+
+                            let lhs = vir::Expr::Field(std::boxed::Box::from(target_place.clone()), vir::Field::new("val_bool", vir::Type::Bool), vir::Position::default());
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]:
```suggestion
                            let lhs = vir::Expr::field(target_place.clone(), vir::Field::new("val_bool", vir::Type::Bool));
```
(I also think we have something in encoder for fields so they are not created directly like this?)

File: prusti-viper/src/lib.rs [hunk 29]
```diff
@@ -23,6 +23,7 @@ extern crate rustc_index;
 extern crate rustc_ast;
 extern crate rustc_target;
 extern crate rustc_attr;
+extern crate rustc_apfloat;
```
Comment (Aurel300) [inline prusti-viper/src/lib.rs]:
```suggestion
```

File: viper/src/ast_factory/expression.rs [hunk 30]
```diff
@@ -12,8 +12,7 @@ use ast_factory::structs::Trigger;
 use ast_factory::structs::Type;
 use ast_factory::AstFactory;
 use jni::objects::JObject;
-use viper_sys::wrappers::viper::silver::ast;
-
+use viper_sys::wrappers::viper::silver::ast::{self,};
```
Comment (Aurel300) [inline viper/src/ast_factory/expression.rs]:
```suggestion
use viper_sys::wrappers::viper::silver::ast;
```

File: viper/src/ast_factory/expression.rs [hunk 31]
```diff
@@ -524,7 +637,7 @@ impl<'a> AstFactory<'a> {
         Expr::new(obj)
     }
      */
-    pub fn domain_func_app(
+     pub fn domain_func_app(
```
Comment (Aurel300) [inline viper/src/ast_factory/expression.rs:609]:
```suggestion
    pub fn domain_func_app(
```

File: viper/src/ast_utils.rs [hunk 32]
```diff
@@ -35,15 +35,18 @@ impl<'a> AstUtils<'a> {
...
+        
+        // let s = self.jni.get_string(
+        //     self.jni.unwrap_result(
+        //         fast_pretty_printer_wrapper.call_pretty(
+        //             self.jni
+        //                 .unwrap_result(fast_pretty_printer_wrapper.singleton()),
+        //             program.to_jobject(),
+        //         ),
+        //     ),
+        // );
```
Comment (Aurel300) [inline viper/src/ast_utils.rs]:
```suggestion
        let s = self.jni.get_string(
            self.jni.unwrap_result(
                fast_pretty_printer_wrapper.call_pretty(
                    self.jni
                        .unwrap_result(fast_pretty_printer_wrapper.singleton()),
                    program.to_jobject(),
                ),
            ),
        );
```
What are these changes…

File: prusti-viper/src/encoder/mir_encoder/mod.rs [hunk 33]
```diff
@@ -13,6 +13,7 @@ use crate::encoder::errors::{
     SpannedEncodingResult, EncodingResult
 };
 use crate::encoder::Encoder;
+use num_traits::Float;
```
Comment (Aurel300) [inline prusti-viper/src/encoder/mir_encoder/mod.rs]: Is this used?

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 34]
```diff
@@ -57,6 +58,7 @@ use rustc_index::vec::Idx;
 // use std;
 use std::collections::HashMap;
 use std::collections::HashSet;
+use std::str::FromStr;
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]: Unused?

File: viper/tests/floats_test_program.rs [hunk 35]
```diff
@@ -0,0 +1,98 @@
+use std::fs::File;
```
Comment (Aurel300) [inline viper/tests/floats_test_program.rs]: Is this a new file? Why do we need to test Viper from Prusti?

Comment (vakaras) [reply viper/tests/floats_test_program.rs]: Viper tests for floats are very minimal and API is still unstable. Therefore, I think it is a good IDE to have a test that checks the API calls.

File: viper/tests/floats_test_program.rs [hunk 36]
```diff
@@ -0,0 +1,98 @@
...
+        &[val_float],
+        &[],
+        &[pred],
+        &[id],
+    );
+
+    let verifier = verification_context.new_verifier(viper::VerificationBackend::Silicon, None);
+    let verification_result = verifier.verify(program);
+    
+    let pretty_printed = ast_utils.pretty_print(program);
```
Comment (vakaras) [inline viper/tests/floats_test_program.rs]: Please remove the debug code.

File: viper/tests/floats_test_program.rs [hunk 37]
```diff
@@ -0,0 +1,98 @@
...
+
+#[test]
+fn test_program() {
+    env_logger::init();
+
+    let verification_context: VerificationContext = VIPER.new_verification_context();
+    let ast = verification_context.new_ast_factory();
+    let ast_utils = verification_context.new_ast_utils();
+
+
```
Comment (vakaras) [inline viper/tests/floats_test_program.rs]: Please remove the unnecessary space.

File: prusti-viper/src/encoder/mir_encoder/mod.rs [hunk 38]
```diff
@@ -484,43 +485,93 @@ impl<'p, 'v: 'p, 'tcx: 'v> MirEncoder<'p, 'v, 'tcx> {
...
+                    mir::BinOp::Rem => vir::Expr::rem(left, right),
+                    mir::BinOp::Div => vir::Expr::div(left, right),
+                    mir::BinOp::Mul => vir::Expr::mul(left, right),
+                    mir::BinOp::BitAnd if is_bool => vir::Expr::and(left, right),
+                    mir::BinOp::BitOr if is_bool => vir::Expr::or(left, right),
+                    mir::BinOp::BitXor if is_bool => vir::Expr::xor(left, right),
+                    mir::BinOp::BitAnd => vir::Expr::bit_and(left, right),
+                    mir::BinOp::BitOr => vir::Expr::bit_or(left, right),
+                    mir::BinOp::BitXor => vir::Expr::bit_xor(left, right),
+                    mir::BinOp::Shl => unimplemented!(),
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir_encoder/mod.rs]: Please remove all `unimplemented!`. If something is not implemented, it should report a meaningful `Err(EncodingError::unsupported(...))`.

File: prusti-common/src/vir/ast/expr.rs (hunk 6 above)
Comment (vakaras) [reply prusti-common/src/vir/ast/expr.rs:106]: That is how the Rust compiler stores them.
